File: Cpp/Reverse_Linked_List_II.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int left, int right) {
        if (head == NULL || left == right)
            return head;
        ListNode* dummy = new ListNode(0);
        dummy->next = head;
        ListNode* pre = dummy;
        for (int i = 0; i < left - 1; i++) {
            pre = pre->next;
        }
        ListNode* curr = pre->next;
        ListNode* then = curr->next;
        for (int i = 0; i < right - left; i++) {
            curr->next = then->next;
            then->next = pre->next;
            pre->next = then;
            then = curr->next;
        }
        ListNode* newHead = dummy->next;
        delete dummy;
        return newHead;
    }
};

File: Rust/Reverse_Linked_List_II.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
//
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn reverse_between(
        head: Option<Box<ListNode>>,
        left: i32,
        right: i32,
    ) -> Option<Box<ListNode>> {
        if left == right {
            return head;
        }
        let mut dummy = Box::new(ListNode { val: 0, next: head });
        let mut pre = &mut dummy;
        for _ in 0..(left - 1) {
            pre = pre.next.as_mut().unwrap();
        }
        let mut curr = pre.next.take().unwrap();
        for _ in 0..(right - left) {
            let mut then = curr.next.take().unwrap();
            curr.next = then.next.take();
            then.next = pre.next.take();
            pre.next = Some(then);
        }
        let mut tail = pre.next.as_mut().unwrap();
        while tail.next.is_some() {
            tail = tail.next.as_mut().unwrap();
        }
        tail.next = Some(curr);

        dummy.next
    }
}
